### PR TITLE
[Grappler] Improve HoistCWiseUnaryChainsStage to support Reshape

### PR DESCRIPTION
**Pull request by xinan-jiang**
It continues the issue #40542

### CODE REVIEW

**Review by ezhulenev — APPROVED** [review]

**Review by rthadur — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

File: tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc
Comment (rmlarsen) [inline tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc:1888]: Check return status here. you can use the TF_RETURN_IF_ERROR macro.

**Comment by gbaned** [comment]
@xinan-jiang  Can you please check @rmlarsen's comments and keep us posted ? Thanks!

File: tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc
Comment (xinan-jiang) [reply tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc]: done

**Review by rthadur — APPROVED** [review]

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
This appears to produce an invalid graph in the test `tensorflow/python/distribute:cross_device_ops_test`. Please take a look. 


Traceback (most recent call last):
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/py/absl/testing/parameterized.py", line 282, in bound_param_test
    return test_method(self, **testcase_params)
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/framework/test_combinations.py", line 367, in decorated
    execute_test_method()
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/framework/test_combinations.py", line 350, in execute_test_method
    test_method(**kwargs_to_pass)
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/distribute/combinations.py", line 435, in decorator
    test_method(self, **kwargs)
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/distribute/cross_device_ops_test.py", line 454, in testBatchAllReduceDense
    self.batch_reduce_and_verify(inputs, expect, options)
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/distribute/cross_device_ops_test.py", line 301, in batch_reduce_and_verify
    get_global_mpr(options.num_processes).run(replica_fn)
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/distribute/multi_process_runner.py", line 980, in run
    six.reraise(*process_status.exc_info)
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/py/six/__init__.py", line 703, in reraise
    raise value
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/distribute/multi_process_runner.py", line 1037, in _run_contained
    return_value = fn(*args, **kwargs)
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/distribute/cross_device_ops_test.py", line 298, in replica_fn
    got = def_function.function(batch_reduce_fn)()
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/eager/def_function.py", line 784, in __call__
    result = self._call(*args, **kwds)
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/eager/def_function.py", line 851, in _call
    filtered_flat_args, self._concrete_stateful_fn.captured_inputs)  # pylint: disable=protected-access
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/eager/function.py", line 1948, in _call_flat
    ctx, args, cancellation_manager=cancellation_manager))
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/eager/function.py", line 561, in call
    ctx=ctx)
  File "/build/work/54d02ca2125b6ddae3c411e2d47c84f01697/google3/runfiles/google3/third_party/tensorflow/python/eager/execute.py", line 60, in quick_execute
    inputs, attrs, num_outputs)
**google3.third_party.tensorflow.python.framework.errors_impl.InvalidArgumentError: Incomplete graph, missing 1 inputs for ArithmeticOptimizer/_Reshape_2_split [Op:__inference_batch_reduce_fn_8924]**

**Comment by gbaned** [comment]
@xinan-jiang  Can you please check @rmlarsen's comments and keep us posted ? Thanks!

**Comment by xinan-jiang** [comment]
Sorry, I can not get the error on my local. It seems a multi-device case.

**Comment by gbaned** [comment]
@rmlarsen  Can you please take a look on above comments from @xinan-jiang. Thanks!

**Comment by gbaned** [comment]
@rmlarsen  Any update on this PR? Please. Thanks!

**Review by rmlarsen — APPROVED** [review]

**Comment by xinan-jiang** [comment]
@rmlarsen  It is a bugfix and modified UT for this bug.

**Comment by xinan-jiang** [comment]
@rmlarsen @ezhulenev @rthadur  could you help to review the new fix.

**Review by ezhulenev — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by gbaned** [comment]
@xinan-jiang Can you please resolve conflicts? Thanks!

**Comment by rmlarsen** [comment]
@xinan-jiang sorry for the long delay. You need to rebase. After rebasing and adding TF_RETURN_IF_ERROR in line ~1818, your new tests do not pass anymore. Mind taking a look.

**Comment by xinan-jiang** [comment]
> @xinan-jiang sorry for the long delay. You need to rebase. After rebasing and adding TF_RETURN_IF_ERROR in line ~1818, your new tests do not pass anymore. Mind taking a look.

Test fail because of the previous commit. Now it could pass.

**Review by rmlarsen — APPROVED** [review]
LGTM

**Comment by xinan-jiang** [comment]
add fix for keras-utils-tests

**Comment by xinan-jiang** [comment]
@rmlarsen @ezhulenev help to review

**Review by rmlarsen — APPROVED** [review]

**Comment by xinan-jiang** [comment]
@gbaned, What blocks this PR ?

**Comment by gbaned** [comment]
> @gbaned, What blocks this PR ?

@xinan-jiang  Internal errors are appearing with this PR, @rmlarsen is working on this. Thanks!

**Comment by rmlarsen** [comment]
This still appears to break the test in tensorflow/contrib/recurrent:functional_rnn_test. Please take a look.

**Comment by rmlarsen** [comment]
@xinan-jiang the issue is that the change appears to break the test in tensorflow/contrib/recurrent:functional_rnn_test. The contrib module was removed in TF2, but we still test against it internally for legacy users. You could try to test your change against it using TF release 1.15.

**Comment by gbaned** [comment]
@xinan-jiang Can you please check @rmlarsen's comment and resolve conflicts? Thanks!

**Comment by xinan-jiang** [comment]
@gbaned  I am working on this.

**Comment by xinan-jiang** [comment]
@rmlarsen @gbaned  It fails with tf1.x "Concat" op, which is instead of "ConcatV2" in tf2.x. Now it could pass with "Concat" op.

**Comment by gbaned** [comment]
@xinan-jiang  Can you please resolve conflicts? Thanks!

**Comment by xinan-jiang** [comment]
@rmlarsen please help to review my resolve for confilct with e98fc97bee12655f93627604515a7140eea1cfe0

**Comment by xinan-jiang** [comment]
@rmlarsen @gbaned help to review.

**Comment by xinan-jiang** [comment]
@ezhulenev @rthadur help to review it.

**Review by rmlarsen — APPROVED** [review]

**Comment by xinan-jiang** [comment]
@gbaned I could not see the details of the fail test. Does it block the merge work?

**Comment by gbaned** [comment]
> @gbaned I could not see the details of the fail test. Does it block the merge work?

@xinan-jiang This is waiting for internal approval, once done it will be processed to merge.  Thanks!

**Comment by rmlarsen** [comment]
I'm afraid this had to be rolled back since it broke tests in the lingvo project. It still seems like the rewrite violates some graph invariants. 

```google3.third_party.tensorflow.python.framework.errors_impl.InvalidArgumentError: Input to reshape is a tensor with 2688000 values, but the requested shape has 640000
	 [[node fprop/test_dtmodel/tower_0_0/Reshape_15 (defined at /third_party/py/lingvo/core/layers.py:2501) ]]```

**Pull request merged**
